cpp/util/etcd.cc

```diff
@@ -597,24 +624,36 @@ void EtcdClient::WatchRequestDone(WatchState* state, GenericResponse* gen_resp,
   // these objects.
   unique_ptr<GenericResponse> gen_resp_deleter(gen_resp);
 
-  // TODO(alcutter): doing this here works around some etcd 401 errors, but in
-  // the case of sustained high qps we could miss updates entirely (e.g. if
-  // this new index is already past the 1000 entry horizon by the time we make
-  // the new watch request.)  One way to address this might be to have the
-  // watcher re-do an "initial" get on the target, and, in the case of
-  // directory watches, maintain a set of known keys so that it can synthesise
-  // 'delete' updates.
-  if (child_task->status().ok()) {
-    VLOG(2) << "etcd_index: " << gen_resp->etcd_index;
-    CHECK_LE(state->highest_index_seen_, gen_resp->etcd_index);
-    state->highest_index_seen_ = gen_resp->etcd_index;
-  }
-
   if (state->task_->CancelRequested()) {
     state->task_->Return(Status::CANCELLED);
     return;
   }
 
+  // Handle when the request index is too old, we have to restart the
+  // watch logic.
+  if (child_task->status().CanonicalCode() == util::error::ABORTED &&
+      gen_resp->etcd_index >= 0) {
+    VLOG(1) << "etcd index: " << gen_resp->etcd_index;
+    state->highest_index_seen_ =
+        max(state->highest_index_seen_, gen_resp->etcd_index);
+
+    if (KeyIsDirectory(state->key_)) {
+      GetAll(state->key_, bind(&EtcdClient::WatchInitialGetAllDone, this,
+                               state, _1, _2, _3));
+    } else {
+      Get(state->key_,
+          bind(&EtcdClient::WatchInitialGetDone, this, state, _1, _2, _3));
+    }
+
+    return;
+  }
+
+  if (child_task->status().ok()) {
+    VLOG(2) << "X-Etcd-Index: " << gen_resp->etcd_index;
+    CHECK_LE(state->highest_index_seen_, gen_resp->etcd_index);
```

Comment:
We've seen etcd give us a crazy low `X-Etcd-Index` when the real is much higher (in fact, more than the watch window of 1000 entries higher!), is it worth changing this to a LOG(WARNING) ?